src/Lecture2.hs

```diff
@@ -74,7 +78,17 @@ return the removed element.
 >>> removeAt 10 [1 .. 5]
 (Nothing,[1,2,3,4,5])
 -}
-removeAt = error "TODO"
+removeAt :: Int -> [a] -> (Maybe a, [a])
+removeAt n ls = removeAt' n ls []
+    where
+      removeAt' :: Int -> [a] -> [a] -> (Maybe a, [a])
+      removeAt' 0 [] preLs = (Nothing, preLs)
+      removeAt' _ [] preLs = (Nothing, preLs)
+      removeAt' n (x:xs) preLs
+        | n == 0 = (Just x, preLs ++ xs)
+        | n < 0 = (Nothing, x:xs)
+        | otherwise = removeAt' (n - 1) xs (preLs ++ [x])
```

Comment:
It's more efficient to add to the beginning of the list. When you append to the end, it's usually becomes inefficient in the end

Reply:
But how can I keep the order of the original list?
Should I add the beginning of the list for each recursive and return the reversed list at the end of the recursive call?
BTW, is there any way to measure the performance of these different syntaxes?